integrations/kube-agent-updater/pkg/maintenance/basichttp.go

```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package maintenance
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/exp/slices"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/gravitational/teleport/integrations/kube-agent-updater/pkg/cache"
+)
+
+const (
+	maintenancePath = "emergency"
+	httpTimeout     = 10 * time.Second
+	cacheDuration   = time.Minute
+)
+
+var (
+	truthy = []string{"true", "True", "yes", "Yes"}
+	falsy  = []string{"false", "False", "no", "No"}
+)
+
+// basicHTTPMaintenanceClient retrieves the version from an HTTP endpoint
+// it should not be invoked immediately and must be wrapped in a cache layer in
+// order to avoid spamming the version server in case of reconciliation errors.
+// use BasicHTTPMaintenanceTrigger if you need to get a version.
+type basicHTTPMaintenanceClient struct {
+	baseURL *url.URL
+	client  *http.Client
+}
+
+// Get sends an HTTP GET request and returns the version prefixed by "v".
+// It expects the endpoint to be unauthenticated, return 200 and the response
+// body to contain a valid semver tag without the "v".
+func (b *basicHTTPMaintenanceClient) Get(ctx context.Context) (string, error) {
+	versionURL := b.baseURL.JoinPath(maintenancePath)
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, versionURL.String(), nil)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	res, err := b.client.Do(req)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return "", trace.Errorf("non-200 status code received: '%d'", res.StatusCode)
+	}
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	payload := strings.TrimSpace(string(body))
+	// Validating early that the payload can be converted to a boolean allows to
+	// gracefully catch connectivity error causes by mitm infratructure such as
+	// corporate proxies.
+	_, err = stringToBool(payload)
```

Comment:
Isn't it better for the auth to return the maintenance window instead of `true`, `no`?

Reply:
The maintenance window is passed by the teleport agent in a configmap. This basic HTTP maintenance trigger is our global emergency trigger in case we break all agents, cannot wait for the next maintenance window and the agents are not able anymore to report the maintenance window.

This endpoint is described here: https://github.com/gravitational/teleport/blob/master/rfd/0109-cloud-agent-upgrades.md#version-discovery-endpoint

Reply:
The version and maintenance are not returned by the auth but by a cross-tenant s3 bucket.